packages/backend/src/open_payments/client/service.ts

```diff
@@ -68,3 +164,85 @@ export async function getIncomingPayment(
     return undefined
   }
 }
+
+async function getIncomingPayment(
+  deps: ServiceDependencies,
+  url: string
+): Promise<IncomingPaymentJSON | undefined> {
+  const requestHeaders = {
+    Authorization: `GNAP ${deps.accessToken}`,
+    'Content-Type': 'application/json'
+  }
+  try {
+    const { status, data } = await axios.get(url, {
+      headers: requestHeaders,
+      timeout: REQUEST_TIMEOUT,
+      validateStatus: (status) => status === 200
+    })
+    if (
+      !deps.validateIncomingPayment({
+        status,
+        body: data
+      }) ||
+      !isValidIncomingPayment(data)
+    ) {
+      throw new Error('unreachable')
+    }
+    return data
+  } catch (_) {
+    return undefined
+  }
+}
+
+const CONNECTION_URL_REGEX = /\/connections\/(.){36}$/
+const INCOMING_PAYMENT_URL_REGEX = /\/incoming-payments\/(.){36}$/
+
+async function getReceiver(
+  deps: ServiceDependencies,
+  url: string
+): Promise<Receiver | undefined> {
+  if (url.match(CONNECTION_URL_REGEX)) {
+    const connection = await getConnection(deps, url)
+    if (connection) {
+      return Receiver.fromConnection(connection)
+    }
+  } else if (url.match(INCOMING_PAYMENT_URL_REGEX)) {
+    const incomingPayment = await getIncomingPayment(deps, url)
+    if (incomingPayment) {
+      return Receiver.fromIncomingPayment(incomingPayment)
+    }
+  }
+}
+
+// Validate referential integrity, which cannot be represented in OpenAPI
+function isValidIncomingPayment(
+  payment: IncomingPaymentJSON
+): payment is IncomingPaymentJSON {
+  if (payment.incomingAmount) {
+    const incomingAmount = parseAmount(payment.incomingAmount)
+    const receivedAmount = parseAmount(payment.receivedAmount)
+    if (
+      incomingAmount.assetCode !== receivedAmount.assetCode ||
+      incomingAmount.assetScale !== receivedAmount.assetScale
+    ) {
+      return false
+    }
+    if (incomingAmount.value < receivedAmount.value) {
+      return false
+    }
+    if (incomingAmount.value === receivedAmount.value && !payment.completed) {
+      return false
+    }
```

Comment:
Wouldn't that be a server error because it should always complete the payment once the full amount has been received?

Reply:
Similar to https://github.com/interledger/rafiki/pull/603#discussion_r974666108
This check is in the context of Rafiki as the client of an arbitrary Open Payments server.
A correctly implemented server shouldn't ever return an incoming payment where `receivedAmount` matches `incomingAmount` but `completed` is `false` (it might return a 500 / server error instead), but the client has to validate the response.